### PR TITLE
Update documentation guidelines for contribution content

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

- Closes https://github.com/apache/datafusion/issues/12357
- related to https://github.com/apache/datafusion/issues/13648

## Rationale for this change

Given recent discussions on improving DataFusion stability (see https://github.com/apache/datafusion/issues/13648) I think it is time to document the criteria to add new features to DataFusion to help ensure everyone's expectations are aligned

## What changes are included in this PR?

1. Add summary of conclusion of https://github.com/apache/datafusion/issues/12357 to the docs

## Are these changes tested?
CI

### CODE REVIEW

**Comment by alamb** [comment]
Rather than state some hard rule, I went with the approach of guidance on what would need more / less discussion prior to acceptance. I felt this would give us flexibility but still given potential contributors guidance

File: docs/source/contributor-guide/index.md [hunk 1]
```diff
@@ -72,13 +72,35 @@ recently active community members in the ticket to get their attention.
...
+start initial experience.
+
+With that in mind, contributions that meet the following criteria are more likely
+to be accepted:
+
+1. Bug fixes for existing features
+2. Test coverage for existing features
+3. Documentation improvements / examples
+4. Performance improvements to existing features (with benchmarks)
+5. "Small" functional improvements to existing features
```
Comment (comphead) [inline docs/source/contributor-guide/index.md:93]:
```suggestion
5. "Small" functional improvements to existing features
```
Non breaking improvements?

File: docs/source/contributor-guide/index.md [hunk 2]
```diff
@@ -72,13 +72,35 @@ recently active community members in the ticket to get their attention.
...
+6. Additional APIs for extending DataFusion's capabilities
+
+Contributions that likely require discussion prior to acceptance include:
```
Comment (comphead) [inline docs/source/contributor-guide/index.md:96]: We probably should note the preferred way to discuss, GH Discussion/ASF slack, etc

File: docs/source/contributor-guide/index.md [hunk 3]
```diff
@@ -72,13 +72,35 @@ recently active community members in the ticket to get their attention.
...
+
+1. New functionality that is part of the "standard sql"
+2. New functions that aren't part of the "standard sql"
```
Comment (comphead) [inline docs/source/contributor-guide/index.md:99]: Not sure I'm following, 🤦  looks like every related to "standard sql" requires discussion?

File: docs/source/contributor-guide/index.md [hunk 4]
```diff
@@ -72,13 +72,35 @@ recently active community members in the ticket to get their attention.
...
+focused as possible, while still satisfying the [design goal] of an easy to
+start initial experience.
+
+With that in mind, contributions that meet the following criteria are more likely
+to be accepted:
+
+1. Bug fixes for existing features
+2. Test coverage for existing features
+3. Documentation improvements / examples
+4. Performance improvements to existing features (with benchmarks)
```
Comment (comphead) [inline docs/source/contributor-guide/index.md:92]: 👍 love benchmarks

File: docs/source/contributor-guide/index.md [hunk 5]
```diff
@@ -72,13 +72,35 @@ recently active community members in the ticket to get their attention.
...
+5. "Small" functional improvements to existing features
+6. Additional APIs for extending DataFusion's capabilities
```
Comment (comphead) [inline docs/source/contributor-guide/index.md:94]:
```suggestion
6. Additional APIs for extending DataFusion's capabilities
7. CI improvements
```

File: docs/source/contributor-guide/index.md [hunk 6]
```diff
@@ -72,13 +72,35 @@ recently active community members in the ticket to get their attention.
...
 ### What Features are Good Fits for DataFusion?
 
-DataFusion is designed to highly extensible, and many features can be implemented
-as extensions without changing the core of DataFusion.
-
-We are [working on criteria for what features are good fits for DataFusion], and
-will update this section when we have more to share.
-
-[working on criteria for what features are good fits for datafusion]: https://github.com/apache/datafusion/issues/12357
+DataFusion is designed to highly extensible, and many features can be
```
Comment (findepi) [inline docs/source/contributor-guide/index.md:75]:
```suggestion
DataFusion is designed to be highly extensible, and many features can be
```

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
BTW We should strive to make datafusion easier to extend. This is not only about particular APIs but also about internal design. We may be principled or not about what kind of SQL variant we provide out of the box, but we should assume extending systems provide something else (eg Ballista speaks Spark SQL, which is different than Datafusion SQL). This is obviously a teaser to longer discussion (https://github.com/apache/datafusion/issues/12723). 

I approved this PR based on my understand that the wording proposed here doesn't preclude -- and actually encourages -- ground work features serving the extensibility story (the issue linked above, https://github.com/apache/datafusion/issues/12604, etc.)

**Comment by alamb** [comment]
> BTW We should strive to make datafusion easier to extend. 

I 100% agree

> I approved this PR based on my understand that the wording proposed here doesn't preclude -- and actually encourages -- ground work features serving the extensibility story (the issue linked above, https://github.com/apache/datafusion/issues/12604, etc.)

Yes, this was my intention and I think it fits directly [with the stated design goals](https://docs.rs/datafusion/latest/datafusion/index.html#design-goals) (aka customizable everything):

1. Work “out of the box”: Provide a very fast, world class query engine with minimal setup or required configuration.
2. Customizable everything: All behavior should be customizable by implementing traits.
3. Architecturally boring 🥱: Follow industrial best practice rather than trying cutting edge, but unproven, techniques.

File: docs/source/contributor-guide/index.md (hunk 1 above)
Comment (alamb) [reply docs/source/contributor-guide/index.md]: That is a good qualification -- non breaking changes are certainly more likely to get approved in my experience. I tried to clarify

File: docs/source/contributor-guide/index.md (hunk 3 above)
Comment (alamb) [reply docs/source/contributor-guide/index.md]: I changed the wording like this so nothing was "required" but to give a hint that major new features are likely to involve more discussion

```
Contributions that will likely involve more discussion (see Discussing New 
Features above) prior to acceptance include:

1. Major new functionality (even if it is part of the "standard sql")
2. New functions, especially if they aren't part of "standard sql"
3. New data sources (e.g. support for Apache ORC)
```

**Comment by comphead** [comment]
Should we let it hang for day-two to get other reviews?

**Comment by alamb** [comment]
> Should we let it hang for day-two to get other reviews?

Yes I think we should leave this open for several more days to make sure anyone who wants a chance to comment can do so. Obviously we can always update the wordiing with subsequent PRs too, but I don't think there is any reason to rush this one in

**Review by Weijun-H — APPROVED** [review]
👍

**Review by phillipleblanc — APPROVED** [review]
Looks great, thanks!

File: docs/source/contributor-guide/index.md [hunk 7]
```diff
@@ -70,15 +70,39 @@ line PR.
...
+4. Performance improvements to existing features (with benchmarks)
+5. "Small" functional improvements to existing features (if they don't change existing behavior)
+6. Additional APIs for extending DataFusion's capabilities
+7. CI improvements
+
+Contributions that will likely involve more discussion (see Discussing New
+Features above) prior to acceptance include:
+
+1. Major new functionality (even if it is part of the "standard sql")
+2. New functions, especially if they aren't part of "standard sql"
```
Comment (findepi) [inline docs/source/contributor-guide/index.md:101]:
```suggestion
1. Major new functionality (even if it is part of the "standard SQL")
2. New functions, especially if they aren't part of "standard SQL"
```

**Review by findepi — APPROVED** [review]

**Review by xudong963 — APPROVED** [review]
I'm late, make sense!

**Comment by Omega359** [comment]
LGTM

**Comment by alamb** [comment]
Thanks for all the comments so far. I plan to merge this tomorrow, Monday Dec 16 unless anyone else would like time to comment

**Comment by alamb** [comment]
🚀 📖

**Pull request merged**